SL/Controller/ZUGFeRD.pm

```diff
@@ -77,72 +97,174 @@ sub action_import_zugferd {
+    $::lxdebug->message(LXDebug::DEBUG1(), "Could not extract ZUGFeRD data, error message: " . $res{'message'});
+    die(t8("Could not extract Factur-X/ZUGFeRD data, data and error message:") . " $res{'message'}");
   }
 
-  my $type = $dom->findnodes('//rsm:ExchangedDocument/ram:TypeCode')->string_value;
+  $parser = $res{'invoice_xml'};
+
+  # Shouldn't be neccessary with SL::XMLInvoice doing the heavy lifting, but
+  # let's grab it, just in case.
+  $dom  = $parser->{dom};
+
+  my %metadata = %{$parser->metadata};
+  my @items = @{$parser->items};
+
+  my $iban = $metadata{'iban'};
+  my $invnumber = $metadata{'invnumber'};
+
+  if ( ! ($metadata{'ustid'} or $metadata{'taxnumber'}) ) {
+    die t8("Cannot process this invoice: neither VAT ID nor tax ID present.");
+    }
+
+  $vendor = find_vendor($metadata{'ustid'}, $metadata{'taxnumber'});
+
+  die t8("Please add a valid VAT ID or tax number for this vendor: #1", $metadata{'vendor_name'}) unless $vendor;
 
-  my $dd   = $dom->findnodes('//ram:ApplicableHeaderTradeSettlement' .
-                             '/ram:SpecifiedTradeSettlementPaymentMeans/ram:TypeCode')->string_value;
-  my $direct_debit = $dd == 59 ? 1 : 0;
 
-  my $iban = $dom->findnodes('//ram:ApplicableHeaderTradeSettlement/ram:SpecifiedTradeSettlementPaymentMeans' .
-                             '/ram:PayeePartyCreditorFinancialAccount/ram:IBANID')->string_value;
-  my $ibanmessage;
+  # Create a record template for this imported invoice
+  $template_ap = SL::DB::RecordTemplate->new(
+      vendor_id=>$vendor->id,
+  );
+
+  # Check IBAN specified on bill matches the one we've got in
+  # the database for this vendor.
   $ibanmessage = $iban ne $vendor->iban ? "Record IBAN $iban doesn't match vendor IBAN " . $vendor->iban : $iban if $iban;
 
+  # Use invoice creation date as due date if there's no due date
+  $metadata{'duedate'} = $metadata{'transdate'} unless defined $metadata{'duedate'};
+
+  # parse dates to kivi if set/valid
+  foreach my $key ( qw(transdate duedate) ) {
+    next unless defined $metadata{$key};
+    $metadata{$key} =~ s/^\s+|\s+$//g;
+
+    if ($metadata{$key} =~ /^([0-9]{4})-?([0-9]{2})-?([0-9]{2})$/) {
+    $metadata{$key} = DateTime->new(year  => $1,
+                                    month => $2,
+                                    day   => $3)->to_kivitendo;
+    }
+  }
+
+  # Try to fill in AP account to book against
+  my $ap_chart_id = $::instance_conf->get_ap_chart_id;
+
+  unless ( defined $ap_chart_id ) {
+    # If no default account is configured, just use the first AP account found.
+    my $ap_chart = SL::DB::Manager::Chart->get_all(
+      where   => [ link => 'AP' ],
+      sort_by => [ 'accno' ],
+    );
+    $ap_chart_id = ${$ap_chart}[0]->id;
+  }
+
+  my $currency = SL::DB::Manager::Currency->find_by(
+    name => $metadata{'currency'},
+    );
+
+  $template_ap->assign_attributes(
+    template_name       => "Faktur-X/ZUGFeRD/XRechnung Import $vendor->name, $invnumber",
```

Comment:
Das erzeugt Meldungen wie die hier:

> Information: Die Belegvorlage »Faktur-X/ZUGFeRD/XRechnung Import SL::DB::Vendor=HASH(0x562163f181b8)->name, RN=080365« wurde geladen.

Du kannst Objekte mit Funktionsaufruf so nicht direkt in Strinngs interpolieren. Sinnvoller wäre so etwas wie:

```perl
  template_name => $::locale->text('Faktur-X/ZUGFeRD/XRechnung import #1, #2', $vendor->name, $invnumber),
````

Reply:
Oh, da sind die Ruby-Gewohnheiten mit mir durchgegangen. Repariert.